File: 64-minimum-path-sum/solution.py
class Solution:
  def minPathSum(self, grid: list[list[int]]) -> int:
    n = len(grid[0])
    tracker: list[int] = [0 for _ in range(n)]
    for i, row in enumerate(grid):
      for j, num in enumerate(row):
        left_num = tracker[j - 1] if j > 0 else -1
        up_num = tracker[j] if i > 0 else left_num
        if left_num == up_num == -1:
          left_num = up_num = 0
        elif left_num == -1:
          left_num = up_num
        elif up_num == -1:
          up_num = left_num

        tracker[j] = num + min(left_num, up_num)

    return tracker[n - 1]

File: 64-minimum-path-sum/solution.cpp
#include <vector>
#include <algorithm>

class Solution
{
public:
  int minPathSum(std::vector<std::vector<int>> &grid)
  {
    const int n = grid[0].size();
    std::vector<int> tracker(n, 0);
    for (size_t i = 0; i < grid.size(); ++i)
    {
      const auto &row = grid[i];
      for (size_t j = 0; j < row.size(); ++j)
      {
        const auto num = row[j];
        int left_num = j > 0 ? tracker[j - 1] : -1;
        int up_num = i > 0 ? tracker[j] : left_num;
        if (left_num == -1 && up_num == -1)
        {
          left_num = up_num = 0;
        }
        else if (left_num == -1)
        {
          left_num = up_num;
        }
        else if (up_num == -1)
        {
          up_num = left_num;
        }

        tracker[j] = num + std::min(left_num, up_num);
      }
    }

    return tracker[n - 1];
  }
};